src/main/java/com/moujounakki/archimedesfluids/FluidPool.java

```diff
@@ -7,126 +7,178 @@
+
+        for (BlockPos pos : set) {
             FluidState fluidstate = level.getFluidState(pos);
-            int transfer = Math.min(8-fluidstate.getAmount(),amount);
+            int transfer = Math.min(8 - fluidstate.getAmount(), amount);
             amount -= transfer;
-            ((IMixinFlowingFluid)fluid).changeFluid(level,pos,transfer);
+            ((IMixinFlowingFluid) fluid).changeFluid(level, pos, transfer);
         }
+
         return true;
     }
+
     public boolean checkForFluid(int amount) {
         int foundFluid = 0;
-        while(foundFluid < amount) {
+
+        while (foundFluid < amount) {
             BlockPos pos = explore(false);
-            if(pos == null)
+            if (pos == null) {
                 return false;
-            if(banned.contains(pos))
+            }
+
+            if (banned.contains(pos)) {
                 continue;
+            }
+
             FluidState fluidstate = level.getFluidState(pos);
-            if(!fluidstate.getType().isSame(fluid))
+            if (!fluidstate.getType().isSame(fluid)) {
                 continue;
+            }
+
             foundFluid += fluidstate.getAmount();
         }
+
         return true;
     }
+
     public boolean checkForSpace(int amount) {
         int foundSpace = 0;
-        while(foundSpace < amount) {
+
+        while (foundSpace < amount) {
             BlockPos pos = explore(true);
-            if(pos == null)
+            if (pos == null) {
                 return false;
-            if(banned.contains(pos))
+            }
+
+            if (banned.contains(pos)) {
                 continue;
-            if(level.getBlockState(pos).isAir() || canBeFluidlogged(pos)) {
+            }
+
+            if (level.getBlockState(pos).isAir() || canBeFluidlogged(pos)) {
                 foundSpace += 8;
                 continue;
             }
+
             FluidState fluidstate = level.getFluidState(pos);
-            if(!fluidstate.getType().isSame(fluid))
+            if (!fluidstate.getType().isSame(fluid)) {
                 continue;
-            foundSpace += 8-fluidstate.getAmount();
+            }
+
+            foundSpace += 8 - fluidstate.getAmount();
         }
+
         return true;
     }
 
     private BlockPos explore(boolean lookForAir) {
-        if(unexplored.isEmpty())
-            return null;
-        BlockPos pos = unexplored.removeFirst();
-        explored.add(pos);
-        for(Direction direction : Direction.values()) {
-            BlockPos pos1 = pos.relative(direction);
-            if(explored.contains(pos1))
-                continue;
-            boolean checkForAir = !lookForAir || !(level.getBlockState(pos1).isAir() || canBeFluidlogged(pos1));
-            if(checkForAir && !level.getFluidState(pos1).getType().isSame(fluid))
-                continue;
-            unexplored.add(pos.relative(direction));
+        BlockPos pos;
+        while ((pos = unexplored.poll()) != null) {
```

Comment:
Is there a reason why this is `while` and not `if`?

Reply:
Reduces checks by a LOT, allows the queue to be emptied much more efficiently

Reply:
But it doesn't seem to loop...
It returns unconditionally at the end of the loop.

Reply:
Ooh forgot to mention, I was working on a batch processing feature in junction with that change. Functionally though in this version the if is the same, sorry for the confusion.